contracts/vaults/TreasuryVault/RibbonAutocallVault.sol

```diff
@@ -0,0 +1,461 @@
+            _couponState.nextCouponType,
+            _couponState.nextAutocallBarrierPCT,
+            _couponState.nextCouponBarrierPCT
+        );
+
+        require(_autocallSeller != address(0), "!_autocallSeller");
+        require(
+            _observationPeriodFreq > 0 &&
+                _observationPeriodFreq % 1 days == 0 &&
+                _observationPeriodFreq <= period,
+            "!_observationPeriodFreq"
+        );
+
+        putOption.nextOptionType = _optionType;
+        couponState.nextCouponType = _couponState.nextCouponType;
+        couponState.nextAutocallBarrierPCT = _couponState.autocallBarrierPCT;
+        couponState.nextCouponBarrierPCT = _couponState.couponBarrierPCT;
+
+        nextObservationPeriodFreq = _observationPeriodFreq;
+        autocallSeller = _autocallSeller;
+        numTotalObservationPeriods = period / _observationPeriodFreq;
+    }
+
+    /**
+     * @dev Checks if vault autocallable
+     * @return observation timestamp if autocallable, otherwise 0
+     * @return the number of coupons earned
+     * @return last observation to hit coupon barrier
+     */
+    function autocallable()
+        external
+        view
+        returns (
+            uint256,
+            uint256,
+            uint256
+        )
+    {
+        (
+            uint256 autocallTimestamp,
+            uint256 numCouponsEarned,
+            uint256 lastCouponBarrierBreachObservation
+        ) = _autocallable(IOtoken(optionState.currentOption).expiryTimestamp());
+        return (
+            autocallTimestamp,
+            numCouponsEarned,
+            lastCouponBarrierBreachObservation
+        );
+    }
+
+    /**
+     * @notice Sets next option type
+     * @param _optionType is the next option type
+     */
+    function setOptionType(OptionType _optionType) external onlyOwner {
+        putOption.nextOptionType = _optionType;
+        emit OptionTypeSet(_optionType);
+    }
+
+    /**
+     * @notice Sets the new coupon state
+     * @param _couponType is the coupon type
+     * @param _autocallBarrierPCT is the autocall barrier pct
+     * @param _couponBarrierPCT is the coupon barrier pct
+     */
+    function setCouponState(
+        CouponType _couponType,
+        uint256 _autocallBarrierPCT,
+        uint256 _couponBarrierPCT
+    ) external onlyOwner {
+        _verifyCouponState(_couponType, _autocallBarrierPCT, _couponBarrierPCT);
+
+        couponState.nextCouponType = _couponType;
+        couponState.nextAutocallBarrierPCT = _autocallBarrierPCT;
+        couponState.nextCouponBarrierPCT = _couponBarrierPCT;
+
+        emit CouponStateSet(
+            _couponType,
+            _autocallBarrierPCT,
+            _couponBarrierPCT
+        );
+    }
+
+    /**
+     * @notice Sets the new period and observation period frequency
+     * @param _period is the period
+     * @param _observationPeriodFreq is the observation period frequency
+     */
+    function setPeriodAndObservationFrequency(
```

Comment:
Mostly for context: do all observation periods in a given larger period have the same duration?

If so, does it make sense to enforce `_period % _observationPeriodFreq == 0` ?